text/0000-light-dom.md

```diff
@@ -0,0 +1,293 @@
+- `slottable` `renderedCallback`
+
+In native slots, the ordering is based on the order in the _consumer_ component, whereas in synthetic and light DOM slots, it's based on the ordering in the _slotted_ component.
+
+##### Lazy slots
+
+The main reason for this difference is that synthetic shadow slots and light DOM slots are both _lazy_. For example:
+
+```html
+<!-- x/slottable.html -->
+<template lwc:no-shadow>
+  <template if:true={foo}>
+    <slot lwc:no-shadow></slot>
+  </template>
+</template>
+```
+
+```js
+// slottable.js
+import { LightningElement, track } from 'lwc'
+export default class Slottable extends LightningElement {
+  foo = false
+}
+```
+
+```html
+<!-- app.html -->
+<template>
+  <x-slottable>
+    <x-foo></x-foo>
+  </x-slottable>
+</template>
+```
+
+In this case, `<x-foo>` will never render at all, because it's within a `<template if:true>` block where the condition is false. Similarly, the slot contents would not render if there was no receiving slot for a given slot name, or if the target component contained no slots at all.
+
+These are all differences with native shadow slots, which render eagerly. However, these differences align with the current implementation of synthetic shadow slots. This behavior is maintained in light DOM slots for simplicity's sake in the LWC engine, as well as for the performance benefit of lazy slots.
+
+##### Events and other slot features
+
+The `slotchange` event and `::slotted` CSS pseudo-selector are both unsupported for light DOM slots.
+
+In the case of `slotchange`, it's not clear what `event.target` would be. For `::slotted`, it's not really relevant or useful since light DOM slots can be styled the same as any other light DOM content.
+
+#### Styles
+
+Until now styles used in LWC components were scoped to the component thanks to shadow DOM (or synthetic shadow DOM) style scoping. In the light DOM, component styles naturally leak out of the component; LWC doesn't do any style scoping out of the box. Developers are in charge of writing selectors that are specific enough to target the intended element or pseudo-element.
+
+To support the cases where a shadow DOM element composes a light element, light DOM styles are required to be injected to the closest root node. For a given light DOM element, if all the ancestor components are also light DOM components, the component style sheet will be injected in the document `<head>`. Otherwise, if any of the ancestors is a shadow DOM component, the style has to be injected into the closest shadow root. Upon insertion of a light DOM element, LWC does the following steps:
+
+- look up for the closest root node (`Node.prototype.getRootNode()`)
+- insert the stylesheet if not already present:
+  - if the root node is the HTML document, the style sheet is inserted as a `<style>` element in the `<head>`.
+  - if the root node is a shadow root, the stylesheet is inserted as a `<style>` element as the first shadow root child.
```

Comment:
After trying to implement this, I think this algorithm works for native shadow DOM, but not synthetic shadow DOM. That's because synthetic shadow has no concept of style scoping, so it won't work for the following case:

```html
<x-light-outside>
  <p>I should not be styled</p>
</x-light-outside>
<x-shadow>
  #shadow-root
    <style>p { color: green }</style>
    <x-light-inside>
      <p>I should be styled</p>
    </x-light-inside>
</x-shadow>
```

If `x-light-inside` styles itself, those styles will bleed into `x-light-outside`, but only in synthetic shadow.

To work around this, the algorithm would have to be something like:

1. Look up the closest root node
2. Check if that node is in native shadow or synthetic shadow mode (there could be a mix on the page)
3. If native, just inject a `<style>` tag
4. If synthetic, figure out its hostAttribute/shadowAttribute and generate the scoped styles for that. Then inject those styles into the document `<head>`.

Because of synthetic shadow, it's much trickier than it appears at first glance.

Reply:
In fact, it's not even clear to me that this is possible to implement in synthetic shadow. Consider:

```html
<x-shadow x-shadow_shadow-host>
    <style>[x-shadow_shadow-host] p { color: green }</style>
    <x-light-inside>
      <p>I should be styled</p>
    </x-light-inside>
    <x-shadow-inside>
      <p>I should not be styled</p>
    </x-shadow-inside>
</x-shadow>
```

In this example, `<x-light-inside>` should have its `<p>` styled, but `<x-shadow-inside>` should not have its `<p>` styled. But I can't think of a CSS rules that would accomplish this, while also correctly bleeding out to other light DOM elements inside the `<x-shadow>`, but not to light DOM components outside `<x-shadow>`.
```

Reply:
The only alternative that occurs to me is to redefine "global CSS" to mean "entirely global to the whole document." Then we would just always insert bare `<style>`s (no scoping with attributes) into the `<head>`. I'm not sure this is a great long-term plan, though, since in the world of native shadow DOM it's not the ideal solution.

Reply:
Another problem with synthetic shadow: the root container element is not guaranteed to have styles of its own, meaning it might not have an attribute we can even scope to. We could use the tag name, but then it would apply to all containers with that tag name across the whole DOM, which doesn't work if the light DOM element is only inserted into one of them.

Reply:
This may be solveable, but only if every light DOM element has attributes added which reflect its containing shadow root component. Working on a fix, but it's starting to look a bit gnarly.

Reply:
Nevermind, this seems solvable.

Reply:
Yes, this is resolvable by having the same attribute for all light DOM elements within a given shadow scope.

Reply:
Just realized this is quite a bit harder at the root level. Consider this:

```html
<body>
  <x-light>
    <div>should be styled</div>
  </x-light>
  <x-shadow> <!-- synthetic shadow -->
    #shadow-root
      <div>should not be styled</div>
  </x-shadow>
  <div>should be styled</div> <!-- not rendered by LWC -->
</body>
```

if `<x-light>` has the CSS `div { color: green }`, then we need to style the two light-DOM `<div>`s but not the shadow one. Assuming the second `<div>` is not rendered by LWC, and the `<x-shadow>` uses synthetic shadow, then I can't think of a CSS selector to scope this properly. Unless we use MutationObserver to listen for all elements added into the root context and apply attributes to them.

Reply:
Ah wait, just remembered synthetic shadow doesn't protect from light DOM style injection. Nevermind.

Reply:
We need some kind of clarity on how light DOM styles interact with synthetic shadow, at the global level and within synthetic shadow roots.

Reply:
I'm also starting to question whether this algorithm is even necessary. In native shadow DOM, we can insert the `<style>` right alongside the element, it works the same as hoisting it into its closest shadow root.

Reply:
Actually, `:host` is kind of interesting. Imagine this:

```html
<x-shadow>
  <x-light>
    <style>
      :host { background: blue }
    </style>
  </x-light>
</x-shadow>
```

Should `<x-light>` be able to style its shadow container like this? I can't see anything technically _wrong_ with it, although it's a bit weird.

Reply:
I amended the RFC with more details on how I think global styles should work. I consider this thread resolved.